cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformExecutor.java

```diff
@@ -17,91 +17,110 @@
+    trackedTransformDetail = new TransformDetail(transformDetail, trackedTransformation);
+    defaultEmitter = new DefaultEmitter<>(transformDetail.getMetrics());
   }
 
-  public TransformResponse<OUT> runOneIteration(IN input) throws Exception {
-
-    Map<String, Collection<OUT>> errorRecordsMap = new HashMap<>(transforms.size());
+  public TransformResponse runOneIteration(IN input) throws Exception {
+    executeTransfomation(start, ImmutableList.of(input));
 
-    if (transforms.isEmpty()) {
-      return new TransformResponse<>(Lists.newArrayList((OUT) input).iterator(), errorRecordsMap);
+    Map<String, List<Object>> terminalNodeEntriesMap = new HashMap<>();
+    Map<String, List<Object>> emitterEntries = defaultEmitter.getEntriesMap();
+    for (String key : emitterEntries.keySet()) {
+      if (!connectionsMap.containsKey(key)) {
+        // terminal node
+        terminalNodeEntriesMap.put(key, emitterEntries.get(key));
+      }
     }
 
-    TransformDetail transformDetail = transforms.get(0);
-    DefaultEmitter currentEmitter = emitters.get(0);
-    currentEmitter.reset();
-    Transformation transform = transformDetail.getTransformation();
-    transform.transform(input, currentEmitter);
-
-    if (!currentEmitter.getErrors().isEmpty()) {
-      errorRecordsMap.put(transformDetail.getTransformId(), currentEmitter.getErrors());
-    }
+    Map<String, List<InvalidEntry<Object>>> errors = defaultEmitter.getErrors();
+    return new TransformResponse(terminalNodeEntriesMap, errors);
+  }
 
-    DefaultEmitter previousEmitter = currentEmitter;
-    for (int i = 1; i < transforms.size(); i++) {
-      transformDetail = transforms.get(i);
-      transform = transformDetail.getTransformation();
-      currentEmitter = emitters.get(i);
-      for (Object transformedVal : previousEmitter) {
-        transform.transform(transformedVal, currentEmitter);
+  // this will be called with starting stage name
+  // we will get the transformation of it and also check if this is a terminal node
+  // if its not a terminal node
+  //    1) execute the transformation
+  //    2) using the connections map, recursively call the executeTransformation on the next links
+  // else if its a terminal node
+  //    1) if transformation exists, execute the transformation, else go to step-2
+  //    2) return
+  private <T> void executeTransfomation(String stageName, List<T> input) throws Exception {
+    Transformation<T, Object> transformation = trackedTransformDetail.getTransformation(stageName);
+    if (transformation == null) {
+      // could be either source or sink
+      // get the next connections, if they are not empty, its a source
+      List<String> nextStages = connectionsMap.get(stageName);
+      if (nextStages != null) {
+        // source
+        for (String nextStageName : nextStages) {
+          executeTransfomation(nextStageName, input);
+        }
+      } else {
+        // its a sink, add the input to the emitter entries list
```

Comment:
similarly, there shouldn't be a concept of sink here. If you reach a terminal node, you should add the transform results to the output.